# DP: Author: Kazuhiro Inaoka
# DP: Status: in CVS 4.0-branch
# DP: PIE Support
Index: gcc-h8300-hms-3.4.6/gcc/config.gcc
===================================================================
--- gcc-h8300-hms-3.4.6.orig/gcc/config.gcc	2011-06-08 23:54:56.000000000 +0200
+++ gcc-h8300-hms-3.4.6/gcc/config.gcc	2011-06-08 23:55:02.000000000 +0200
@@ -1324,7 +1324,6 @@
 	;;
 m32r-*-linux*)
 	tm_file="dbxelf.h elfos.h svr4.h linux.h ${tm_file} m32r/linux.h"
-	extra_parts="crtbegin.o crtbeginS.o crtend.o crtendS.o"
 	tmake_file="t-slibgcc-elf-ver m32r/t-linux"
 	gnu_ld=yes
 	use_fixproto=yes
@@ -1334,7 +1333,6 @@
  	;;
 m32rle-*-linux*)
 	tm_file="dbxelf.h elfos.h svr4.h linux.h m32r/little.h ${tm_file} m32r/linux.h"
-	extra_parts="crtbegin.o crtbeginS.o crtend.o crtendS.o"
 	tmake_file="t-slibgcc-elf-ver m32r/t-linux"
 	gnu_ld=yes
 	use_fixproto=yes
Index: gcc-h8300-hms-3.4.6/gcc/config/m32r/m32r.h
===================================================================
--- gcc-h8300-hms-3.4.6.orig/gcc/config/m32r/m32r.h	2004-07-07 13:08:37.000000000 +0200
+++ gcc-h8300-hms-3.4.6/gcc/config/m32r/m32r.h	2011-06-08 23:55:02.000000000 +0200
@@ -191,7 +191,7 @@
 
 /* Options to pass on to the assembler.  */
 #undef  ASM_SPEC
-#define ASM_SPEC "%{v} %(asm_cpu) %(relax) %{fpic:-K PIC} %{fPIC:-K PIC}"
+#define ASM_SPEC "%{v} %(asm_cpu) %(relax) %{fpic|fpie:-K PIC} %{fPIC|fPIE:-K PIC}"
 
 #define LINK_SPEC "%{v} %(link_cpu) %(relax)"
 
Index: gcc-h8300-hms-3.4.6/gcc/config/m32r/linux.h
===================================================================
--- gcc-h8300-hms-3.4.6.orig/gcc/config/m32r/linux.h	2011-06-08 23:53:32.000000000 +0200
+++ gcc-h8300-hms-3.4.6/gcc/config/m32r/linux.h	2011-06-08 23:55:02.000000000 +0200
@@ -85,14 +85,20 @@
     %{profile:-lc_p} %{!profile: -lc}}"
 
 #undef  STARTFILE_SPEC
+#if defined HAVE_LD_PIE
+#define STARTFILE_SPEC \
+  "%{!shared: %{pg|p|profile:gcrt1.o%s;pie:Scrt1.o%s;:crt1.o%s}} \
+  crti.o%s %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+#else
 #define STARTFILE_SPEC \
   "%{!shared: \
      %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:crt1.o%s}}}\
    crti.o%s %{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}"
+#endif
 
 #undef  ENDFILE_SPEC
 #define ENDFILE_SPEC \
-  "%{!shared:crtend.o%s} %{shared:crtendS.o%s} crtn.o%s"
+  "%{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
 
 #undef  SUBTARGET_CPP_SPEC
 #define SUBTARGET_CPP_SPEC "\
